### PR TITLE
Jog cpp api fixes

### PR DESCRIPTION
**Pull request by AndyZe**
This cleans up a few things in jog_arm:

- Renames JogCppApi class to JogCppInterface to match the filename

- Fix a crash when the dimensions of the Jacobian changed

- Doxygenize some comments

- Make the inheritance of JogInterfaceBase public. I wanted this so I could call the functions from external C++ code, like this:  `jog_interface_ptr_->changeDriftDimensions(...)`

### CODE REVIEW

File: moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp [hunk 1]
```diff
@@ -751,17 +750,19 @@ bool JogCalcs::addJointIncrements(sensor_msgs::JointState& output, const Eigen::
 
 void JogCalcs::removeDimension(Eigen::MatrixXd& jacobian, Eigen::VectorXd& delta_x, unsigned int row_to_remove)
 {
-  unsigned int num_rows = jacobian.rows() - 1;
-  unsigned int num_cols = jacobian.cols();
+  /*
```
Comment (henningkayser) [inline moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: Why did you remove this?

File: moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp [hunk 2]
```diff
@@ -500,8 +500,7 @@ double JogCalcs::decelerateForSingularity(const Eigen::VectorXd& commanded_veloc
   new_theta += pseudo_inverse_ * delta_x;
   kinematic_state_->setJointGroupPositions(joint_model_group_, new_theta);
 
-  jacobian_ = kinematic_state_->getJacobian(joint_model_group_);
-  Eigen::JacobiSVD<Eigen::MatrixXd> new_svd = Eigen::JacobiSVD<Eigen::MatrixXd>(jacobian_);
+  Eigen::JacobiSVD<Eigen::MatrixXd> new_svd = Eigen::JacobiSVD<Eigen::MatrixXd>(jacobian);
```
Comment (henningkayser) [inline moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]:
```suggestion
  Eigen::JacobiSVD<Eigen::MatrixXd> new_svd(jacobian);
```

File: moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp (hunk 1 above)
Comment (AndyZe) [reply moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: Oops, it was commented for debugging. Will restore.

File: moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp (hunk 2 above)
Comment (AndyZe) [reply moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: done

File: moveit_experimental/moveit_jog_arm/include/moveit_jog_arm/jog_calcs.h [hunk 3]
```diff
@@ -154,8 +154,8 @@ class JogCalcs
 
   JogArmParameters parameters_;
 
-  // For jacobian calculations
-  Eigen::MatrixXd jacobian_, pseudo_inverse_, matrix_s_;
+  // Preallocate for jacobian calculations
```
Comment (jliukkonen) [inline moveit_experimental/moveit_jog_arm/include/moveit_jog_arm/jog_calcs.h]: I would reword comments since the lines below don't really pre-allocate memory but rather cache results that are calculated once(?) Correct me though if I have misunderstood the way Eigen allocates memory for dynamic matrices.

File: moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp [hunk 4]
```diff
@@ -474,23 +474,24 @@ bool JogCalcs::applyVelocityScaling(const JogArmShared& shared_variables, std::m
 
 // Possibly calculate a velocity scaling factor, due to proximity of singularity and direction of motion
 double JogCalcs::decelerateForSingularity(const Eigen::VectorXd& commanded_velocity,
```
Comment (jliukkonen) [inline moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: Could this function be renamed? It calculates a velocity scaling factor but it sounds like it commands the manipulator to decelerate when close to singularity. Maybe `velocityScalingFactorForSingularity` or something like that.

File: moveit_experimental/moveit_jog_arm/include/moveit_jog_arm/jog_calcs.h (hunk 3 above)
Comment (mlautman) [reply moveit_experimental/moveit_jog_arm/include/moveit_jog_arm/jog_calcs.h]: I don't like pre-allocating memory by creating member variables unless there constructor is non-trivial (eg a RobotState variable). It seems like poor design because you never know how they may be used elsewhere and there is no reason why their values should be preserved. Seems like a use case for a static variable instead..

File: moveit_experimental/moveit_jog_arm/include/moveit_jog_arm/jog_cpp_interface.h [hunk 5]
```diff
@@ -44,15 +44,15 @@
 namespace moveit_jog_arm
 {
 /**
-* Class JogCppApi - This class should be instantiated in a new thread
+* Class JogCppInterface - This class should be instantiated in a new thread
 * See cpp_interface_example.cpp
 */
-class JogCppApi : protected JogInterfaceBase
+class JogCppInterface : public JogInterfaceBase
```
Comment (mlautman) [inline moveit_experimental/moveit_jog_arm/include/moveit_jog_arm/jog_cpp_interface.h]: Since this is in moveit_experimental in the master branch, this is okay but in general, it is really bad practice to change public interfaces without deprecating the old one.

File: moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp [hunk 6]
```diff
@@ -321,28 +321,28 @@ bool JogCalcs::cartesianJogCalcs(geometry_msgs::TwistStamped& cmd, JogArmShared&
   Eigen::VectorXd delta_x = scaleCartesianCommand(cmd);
 
   // Convert from cartesian commands to joint commands
-  jacobian_ = kinematic_state_->getJacobian(joint_model_group_);
+  Eigen::MatrixXd jacobian = kinematic_state_->getJacobian(joint_model_group_);
```
Comment (mlautman) [inline moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: If you are worried about performance, you could make this static as well

File: moveit_experimental/moveit_jog_arm/src/jog_cpp_interface.cpp [hunk 7]
```diff
@@ -39,13 +39,13 @@
 
 #include "moveit_jog_arm/jog_cpp_interface.h"
 
-// TODO(davetcoleman): rename JogCppApi to JogCppInterface to match file name
+// TODO(davetcoleman): rename JogCppInterface to JogCppInterface to match file name
```
[the remark this answers is not in the document]
Comment (AndyZe) [reply moveit_experimental/moveit_jog_arm/src/jog_cpp_interface.cpp]: done

File: moveit_experimental/moveit_jog_arm/include/moveit_jog_arm/jog_calcs.h (hunk 3 above)
Comment (AndyZe) [reply moveit_experimental/moveit_jog_arm/include/moveit_jog_arm/jog_calcs.h]: These variables get updated frequently in a while loop (like, 100 times/second or so) so performance kind of matters. I guess I'm not clear how static variables would be better, but I want to know

I'll make them all local variables until I hear back otherwise

File: moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp (hunk 4 above)
Comment (AndyZe) [reply moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: good idea, done

File: moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp (hunk 6 above)
Comment (AndyZe) [reply moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: done

Comment (AndyZe) [reply moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: Strangely, making this static causes a crash. Probably due to some obscure Eigen memory management issue?

Comment (AndyZe) [reply moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: I think the problem is:

> Static variables when used inside function are initialized only once, and then they hold their value even through function calls.

https://www.studytonight.com/cpp/static-keyword.php

Comment (mlautman) [reply moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp]: Sounds like some janky Eigen stuff for sure
